### PR TITLE
Lums/sc 32552/nn descent

### PR DESCRIPTION
**Pull request by lums658**
This PR includes the implementation of two graph-based indexing algorithms: nn-descent and vamana.  The cpp_api branch was also merged into this one, so there is also code in this PR from that branch.

CPP API
======
For C++ library API
* api.h - Type-erased interface between Python and C++, including `Vector`, `FeatureVector`, and `Index`.  The type-erased classes take a URI and read in the corresponding TileDB array.  The underlying type of the `Vector`, `FeatureVector`, or `Index` class is determined by the data type of the TileDB array.
* concepts.h - Initial introduction of C++20 concepts.  A collection of concepts for defining the API to the vector search library.  This allows us to abstract away the format of the data storage (e.g., we don't need to worry about row-major or column-major).
* Use concepts for template parameters in function templates.  Replace type-specific calls in the library (e.g., in query functions, matrix classes) with CPOs.  There will be many small diffs due to this replacement.  The two most important concepts are
  * A `feature_vector` is a contiguous `random_access_range` with the addition of `operator[]` on the container. Moreover, the expression `dimension(v)` must be valid for a type to meet the requirements of `feature_vector` (which can be implemented with a customization point object, or CPO).
  * A `feature_vector_range` is a container of `feature_vector`. The ith `feature_vector` is obtained via `operator[]`. The expressions `dimension(r)` and `num_vectors(r)` must be valid.  `feature_vector_range` is not actually a range (in the sense of `std::ranges`, since we don't provide the required iterators) so we will need to change the name at some point. Or we could consider requiring `feature_vector_range` actually meet the requirements of `std::range` -- in which case it would need to have all the iterator plumbing, etc.  Providing this capability is not too difficult, requiring only a fairly straightforward facade.
* cpos.h - Customization point objects to work in conjunction with the concepts, including
  * `num_vectors()`, `dimension()` to provide abstract size data for set of vectors (rather than `num_rows()`, `num_cols()`)
  * `load()` - abstracts away loading data into a matrix.  Is a nop for a `Matrix`, will load data into a `tdbMatrix`

Further descriptions about the CPP_API PR can be found in PR 131.

Graph-Based Indexing and Search
=========================

* Graph classes
  * adj_list.h - Two adjacency list classes, one that just stores indices and one that stores indices and scores
  * nn-graph.h - An adjacency list class `nn_graph` that keeps the neighbors in a `fixed_min_heap` in order to store only the neighbors with the top-n scores

Graph creation and search algorithms
* nn-descent.h - Implements the nn-descent algorithm, using an `nn_graph`
* vamana.h - Implements a `vamana_index` class that builds an index according to the vamana algorithm as described in the "filtered fresh DiskANN" paper (*not* either the original DiskANN paper nor the fresh DiskANN paper).  
  * The index is built by iterating over the vectors in the data set and processing each vector as an insert into the index.  
  * When a vector is added, a best-first path from the medoid (the vector closest to the middle of the vector data set) is computed. 
  *  The path is then pruned to keep the degree of the vertexes on that path below a given threshold.  
  * The functions for performing the best-first search and the pruning are `greedy_search` and `robust_prune` (which are the names given in the original DiskANN paper).  

Utility and helper functions
* diskann.h - functions for reading DiskANN memory index files

Demo applications
* The `demo_*` programs in the test subdirectory are essentially integration tests that output qualitative data for assessing proper functioning of key algorithms in building the graph-based indexes (such as dumping graph information to disk so that the graphs can be read and plotted by networkx).  Two notebooks are provided for examining data dumped by `unit_vamana` and `vamana`: `plot_unit_vamana.ipynb` and `plot_vamana.ipynb`.

Unit tests
* Unit tests for the new classes and functions are provided in include/test.

CLI drivers
* Two CLI drivers are provided for vamana indexing: src/vamana/index.cc which builds a vamana-based index for a specified TileDB set of vector data, and src/vamana/query.cc that queries a vamana-based index.  As with previous C++ CLI programs, these print extensive logging data.

Benchmark scripts
* src/benchmarks/vamana - diskann.bash and vamana.bash respectively run indexing and querying programs from DiskANN and TileDB on some small-ish datasets (datasets that will fit in memory) and report timings (etc).  DiskANN is only configured to compile and run on x86 Linux, so these scripts need to be run on an appropriate platform.

Cleanup of unused files

*NB*  A number of the unit tests read and write arrays from the local file system.  These are hard-coded to my personal development environment since we are not currently pulling those into a uniform location (this should be fixed at some point so that test arrays are pulled in manually by running a script or automatically as part of cmake).

### CODE REVIEW

File: src/include/algorithm.h [hunk 1]
```diff
@@ -186,6 +194,8 @@ void range_for_each(
            start,
            stop,
            f = std::forward<UnaryFunction>(f)]() mutable {
+            // scoped_timer __{tdb_func__ + std::string{" (lambda)"}, true};
```
Comment (Shelnutt2) [inline src/include/algorithm.h:197]: Remove

File: src/include/concepts.h [hunk 2]
```diff
@@ -36,84 +36,177 @@
...
+//   contiguous -- all the vectors are stored in a single contiguous array
+//   partitioned -- the range is a range of partitions, each of which is a
+//                  contiguous range of vectors
+//   distributed?-- the range is a range of partitions, each of which is a
+//
+// @todo operator()(size_t, size_t) ?
+// ----------------------------------------------------------------------------
+template <class D>
+concept feature_vector_array =
+    // feature_vector<inner_range_t<D>> &&
```
Comment (Shelnutt2) [inline src/include/concepts.h:151]: Remove?

File: src/include/concepts.h [hunk 3]
```diff
@@ -36,84 +36,177 @@
...
+    // std::ranges::random_access_range<D> && /* std::ranges::sized_range<D> &&
+    // */ subscriptable_range<D> && requires(D d, const
+    // std::iter_difference_t<std::ranges::iterator_t<D>> n) {
+    requires(D d, size_t n) {
+  { num_vectors(d) } -> semi_integral;
+  { dimension(d) } -> semi_integral;
+  //{
+  //        d[n]
```
Comment (Shelnutt2) [inline src/include/concepts.h:159]: Remove?

File: src/include/detail/flat/qv.h [hunk 4]
```diff
@@ -147,8 +142,10 @@ auto qv_query_heap(
     int k_nn,
     unsigned nthreads) {
   scoped_timer _{tdb_func__};
+  // load(db);
```
Comment (Shelnutt2) [inline src/include/detail/flat/qv.h:145]: Remove?

File: src/include/detail/flat/qv.h [hunk 5]
```diff
@@ -147,8 +142,10 @@ auto qv_query_heap(
...
 
-  // using feature_type = typename std::remove_reference_t<decltype(db)>::value_type;
+  // using feature_type = typename
+  // std::remove_reference_t<decltype(db)>::value_type;
```
Comment (Shelnutt2) [inline src/include/detail/flat/qv.h:148]: Remove?

File: src/include/detail/flat/qv.h [hunk 6]
```diff
@@ -226,30 +218,31 @@ auto qv_query_heap_tiled(
...
+  load(db);
 
-  // using feature_type = typename std::remove_reference_t<decltype(db)>::value_type;
-  using id_type = typename std::remove_reference_t<decltype(ids)>::value_type;
-  using score_type = float;
-
-  if constexpr (is_loadable_v<decltype(db)>) {
-    db.load();
-  }
+  // debug_slice(db);
```
Comment (Shelnutt2) [inline src/include/detail/flat/qv.h:223]: Remove?

File: src/include/detail/flat/vq.h [hunk 7]
```diff
@@ -184,25 +147,21 @@ auto vq_query_heap_tiled(
...
-    const ID& ids,
-    int k_nn,
-    unsigned nthreads) {
+    T, DB& db, const Q& q, const ID& ids, int k_nn, unsigned nthreads) {
   // @todo Need to get the total number of queries, not just the first block
   // @todo Use Matrix here rather than vector of vectors
 
-  // using feature_type = typename std::remove_reference_t<decltype(db)>::value_type;
+  // using feature_type = typename
+  // std::remove_reference_t<decltype(db)>::value_type;
```
Comment (Shelnutt2) [inline src/include/detail/flat/vq.h:155]: Remove?

File: src/include/detail/flat/vq.h [hunk 8]
```diff
@@ -245,46 +200,37 @@ auto vq_query_heap_tiled(
...
-    const Q& q,
-    const ID& ids,
-    int k_nn,
-    unsigned nthreads) {
+    T, DB& db, const Q& q, const ID& ids, int k_nn, unsigned nthreads) {
   // @todo Need to get the total number of queries, not just the first block
   // @todo Use Matrix here rather than vector of vectors
 
-  // using feature_type = typename std::remove_reference_t<decltype(db)>::value_type;
+  // using feature_type = typename
```
Comment (Shelnutt2) [inline src/include/detail/flat/vq.h:223]: Remove?

**Review by Shelnutt2 — COMMENTED** [review]
Minor cleanup comments, still reviewing the rest.

**Comment by ihnorton** [comment]
#162, #173+

**Pull request closed without merging**